### PR TITLE
2.1.1

### PR DESCRIPTION
**Pull request by MihaiAnton**

### CODE REVIEW

File: src/volume/volume.cpp [hunk 1]
```diff
@@ -144,32 +148,79 @@ float Volume::linearInterpolate(float g0, float g1, float factor)
...
     return 0.0f;
 }
 
-// ======= TODO : IMPLEMENT ========
 // This functions returns the results of a cubic interpolation using 4 values and a factor
 float Volume::cubicInterpolate(float g0, float g1, float g2, float g3, float factor)
 {
-    return 0.0f;
+    float value = pow(factor, 3) * (-0.5 * g0 + 1.5 * g1 - 1.5 * g2 + 0.5 * g3) + pow(factor, 2) * (g0 - 2.5 * g1 + 2 * g2 - 0.5 * g3) + factor * (-0.5 * g0 + 0.5 * g2) + g1;
+    return value;
```
Comment (anujinho) [inline src/volume/volume.cpp:169]: Shouldn't cubic interpolation be based on the weight defined? (This is the derivative based solution to the cubic interpolation equation.)

File: src/render/renderer.cpp [hunk 2]
```diff
@@ -188,12 +188,46 @@ float Renderer::bisectionAccuracy(const Ray& ray, float t0, float t1, float isoV
...
+    float opacity = 1.0f;
+
+    for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, samplePos += increment) {
+        glm::vec tf_value = this->getTFValue(this->m_pVolume->getVoxelInterpolate(samplePos));
+
+        color = opacity * color + (1 - opacity) * glm::vec3(tf_value);
+        opacity = opacity + (1 - opacity) * tf_value[3];
+    }
+
+    return glm::vec4(color / m_pVolume->maximum(), 1);
```
Comment (anujinho) [inline src/render/renderer.cpp:230]: Same issue, doesn't work for back to front compositing with the same logic, but works when the loop is reversed (front to back)

**Review by anujinho — COMMENTED** [review]
done

File: src/volume/volume.cpp [hunk 3]
```diff
@@ -162,44 +165,58 @@ float Volume::weight(float x)
...
-        }
-    }
+    const int x = static_cast<int>(xyCoord.x);
+    const int y = static_cast<int>(xyCoord.y);
+
+    const float fac_x = xyCoord.x - float(x);
+    const float fac_y = xyCoord.y - float(y);
 
-    glm::vec1 result = (dist_x * values) * dist_y;
+    for (int i = 0 - offset; i < 4 - offset; i++) {
```
Comment (Aayush2007) [inline src/volume/volume.cpp:187]: edge conditions can be added here, as code gives error when the image is adjusted.

Comment (MihaiAnton) [reply src/volume/volume.cpp:187]: The values are checked [here](https://github.com/MihaiAnton/volume-visualization/blob/3d93d4cad62d0bb6505b3445aba8f0e1901354e5/src/volume/volume.cpp#L81). Is this the possible error you are referring to?

File: src/volume/volume.cpp (hunk 1 above)
Comment (MihaiAnton) [reply src/volume/volume.cpp:169]: I took the parameters from [here](https://www.paulinternet.nl/?page=bicubic). How did you implement this?

**Pull request merged**